go-controller/pkg/ovn/gateway_init.go

```diff
@@ -304,6 +305,184 @@ func gatewayInit(nodeName string, clusterIPSubnet []*net.IPNet, hostSubnets []*n
+			"stdout: %q, stderr: %q, error: %v", distributedGWPortSwitch, lclNetPortname, lspName, stdout, stderr, err)
+	}
+	// finally add an entry to the OVN SB MAC_Binding table, if not present, to capture the
+	// MAC-IP binding of util.V4DnatAndSnatNextHop address to its MAC. Normally, this will
+	// be learnt and added by the chassis to which the distributed gateway port (DGP) is
+	// bound. However, in our case we don't send any traffic out with the DGP port's IP
+	// as source IP, so that binding will never be learnt and we need to seed it.
+	dnatSnatNextHopMac := util.IPAddrToHWAddr(net.ParseIP(util.V4DnatAndSnatNextHop)).String()
+	stdout, stderr, err = util.RunOVNSbctl("--data=bare", "--no-heading", "--columns=_uuid", "find", "MAC_Binding",
+		"logical_port="+dgpName, fmt.Sprintf(`mac="%s"`, dnatSnatNextHopMac))
+	if err != nil {
+		return fmt.Errorf("failed to check existence of MAC_Binding entry of (%s, %s) for distributed router port %s "+
+			"stderr: %q, error: %v", util.V4DnatAndSnatNextHop, dnatSnatNextHopMac, dgpName, stderr, err)
+	}
+	if stdout != "" {
+		klog.Infof("The MAC_Binding entry of (%s, %s) exitsts on distributed router port %s with uuid %s",
+			util.V4DnatAndSnatNextHop, dnatSnatNextHopMac, dgpName, stdout)
+		return nil
+	}
+
+	datapath, stderr, err := util.RunOVNSbctl("--data=bare", "--no-heading", "--columns=_uuid", "find", "datapath",
+		"external_ids:name="+ovnClusterRouter)
+	if err != nil {
+		return fmt.Errorf("failed to get the datapatah UUID of %s from OVN SB "+
+			"stdout: %q, stderr: %q, error: %v", ovnClusterRouter, datapath, stderr, err)
+	}
+
+	_, stderr, err = util.RunOVNSbctl("create", "mac_binding", "datapath="+datapath, "ip="+util.V4DnatAndSnatNextHop,
+		"logical_port="+dgpName, fmt.Sprintf(`mac="%s"`, dnatSnatNextHopMac))
+	if err != nil {
+		return fmt.Errorf("failed to create a MAC_Binding entry of (%s, %s) for distributed router port %s "+
+			"stderr: %q, error: %v", util.V4DnatAndSnatNextHop, dnatSnatNextHopMac, dgpName, stderr, err)
+	}
+
+	return nil
+}
+
+func addPolicyBasedRoutes(nodeName, mgmtPortIP string, hostIfAddrs []*net.IPNet) error {
+	// embed nodeName as comment so that it is easier to delete these rules later on.
+	// logical router policy doesn't support external_ids to stash metadata
+	matchStr := fmt.Sprintf(`inport == "rtos-%s" && ip4.dst == %s /* %s */`,
+		nodeName, hostIfAddrs[0].IP.String(), nodeName)
+	_, stderr, err := util.RunOVNNbctl("lr-policy-add", ovnClusterRouter, nodeSubnetPolicyPriority, matchStr, "reroute",
+		mgmtPortIP)
+	if err != nil {
+		// lr-policy-add doesn't support --may-exist and what the heck?
+		if !strings.Contains(stderr, "already existed") {
+			return fmt.Errorf("failed to add policy route '%s' for host %q on %s "+
+				"stderr: %s, error: %v", matchStr, nodeName, ovnClusterRouter, stderr, err)
+		}
+	}
+
+	matchStr = fmt.Sprintf("ip4.src == %s && ip4.dst == %s /* %s */",
```

Comment:
The comment is only used as a metadata because router_policy table didn't have external_ids column and it is only present in the master branch of OVN repository.